kernel/thread.c

```diff
@@ -1718,3 +1723,67 @@ int myst_interrupt_thread(myst_thread_t* thread)
 done:
     return ret;
 }
+
+int myst_set_signal_delivery_altstack(
+    myst_thread_t* thread,
+    size_t stack_size,
+    pid_t pid)
+{
+    int ret = 0;
+    void* stack = NULL;
+
+    if (!thread || thread->signal_delivery_altstack ||
+        thread->signal_delivery_altstack_size)
+        ERAISE(-EINVAL);
+
+    if (stack_size % PAGE_SIZE)
+        ERAISE(-EINVAL);
+
+    const int prot = PROT_READ | PROT_WRITE;
+    const int flags = MAP_ANONYMOUS | MAP_PRIVATE;
+
+    stack = (void*)myst_mmap(NULL, stack_size + PAGE_SIZE, prot, flags, -1, 0);
+
+    if ((long)stack < 0)
+        ERAISE(-ENOMEM);
+
+    /* Make the first page as the guard page */
+    ECHECK(myst_mprotect(stack, PAGE_SIZE, PROT_NONE));
+
+    thread->signal_delivery_altstack = stack;
+    thread->signal_delivery_altstack_size = stack_size + PAGE_SIZE;
+
+    /* set pid to the newly allocated stack (now part of the process
+     * mappings) */
+    ECHECK(myst_mman_pids_set(
+        thread->signal_delivery_altstack,
+        thread->signal_delivery_altstack_size,
+        pid));
+
+    myst_tcall_td_set_exception_handler_stack(
+        (void*)thread->target_td,
+        (void*)((uint64_t)stack + PAGE_SIZE),
+        stack_size);
+
+done:
+    return ret;
+}
+
+int myst_clear_signal_delivery_altstack(myst_thread_t* thread)
+{
+    int ret = 0;
+
+    if (!thread)
+        ERAISE(-EINVAL);
+
+    myst_tcall_td_set_exception_handler_stack(
+        (void*)thread->target_td, NULL, 0);
+
+    /* the memory will be freed as part of myst_release_process_mappings */
```

Comment:
Regarding lazy release of memory - In the case a process deregisters its alternate stack via `sigaltstack` or `sigaction`, the process might keep running for a while.

Reply:
Updated.